### ISSUE TITLE
Optionally provide callback for MockedReponse.results

### ISSUE THREAD

**Issue by Samigirma1** [apollographql/apollo-client#8981]
Hello! 😊

I was trying to use the `MockedProvider` to test a connected component. However, there's no way for me to use the variables from the query to generate a dynamic response. Would it be possible to add this feature for the mock provider?

Alternatively, the `mock` prop can be a object with a signature like....
 ```
// This is probably an incorrect typed definition :)
{
  callback: (
      request: DocumentNode | TypedDocumentNode<TData, TVariables>,
      variables?:  TVariables,
      addTypeName?: boolean,
    ) => {
        result?: FetchResult<TData>,
        error?: ApolloError
      },
  delay?: number |
    (
      (request?: DocumentNode | TypedDocumentNode<TData, TVariables>,
        variables?:  { [key: string]: any }) => number)
 } |
  ReadonlyArray<MockedResponse>

``` 
It will use the request and its variables to create a dynamic server response. If provided, it would also delay the response by number of milliseconds specified by the `delay`. If `delay` is a callback, it will calculate the delay amount using the `request` and `variables` for each request.

**Comment by koenpunt** [comment]
A mocked response's result can be a function, would that help?

```tsx
<MockedProvider mocks=[
  {
    request: {
      query: ...,
      variables: { ... }
    },
    result: () => {
      return { data: { ... } };
    }
  }
]>
```

**Comment by Samigirma1** [comment]
Not really. That would still require that I fully enumerate all the requests my components would make in their lifecycle. But in us cases when query variables are dynamic--one example could be a list component with infinite scrolling that varies it's fetchMore size based on screen size--I would have to enumerate an array with all the possible (and reasonable) fetch sizes. I was trying to avoid this by simply using a callback that receives the query and variables/options, and returns the appropriate query object/error.

**Comment by koenpunt** [comment]
And what about the `newData` property of `MockedResponse`?

It's undocumented afaik, but it should allow you to return arbitrary data for any request, without having to define all individual responses. 

Haven't used it myself, but maybe that can help you?

**Comment by Samigirma1** [comment]
That still does not allow query responses to be generated dynamically for each response.

**Comment by jerelmiller** [comment]
Hey all 👋 

Apollo Client 4 did a lot of refinements to `MockLink` and as such `newData` and `result` were combined into the `result` property (https://github.com/apollographql/apollo-client/pull/12530). You can pass a callback to `result` which is called with the `variables` executed by the query.

You can see an example of this in the `MockLink` tests: https://github.com/apollographql/apollo-client/blob/231de9404ec8c0807bca31ff6c2050a64bc2b656/src/testing/core/mocking/__tests__/mockLink.ts#L1609-L1635

(Note: `MockedProvider` uses `MockLink` under the hood, so the `mocks` property works the same.)

Additionally, Apollo Client 4 added support for a callback function for `delay` that lets you determine what delay to be applied to the mock (https://github.com/apollographql/apollo-client/pull/12532). This callback takes the Apollo Link `operation` object as an argument so you can use it to provide dynamic delays based on the query/variables that are executed during the request.

I believe this satisfies both requests here so I'm going to go ahead and close this issue. Please see the [migration guide](https://www.apollographql.com/docs/react/migrating/apollo-client-4-migration) for detailed instructions on upgrading to Apollo Client 4. Thanks!
